pip/qsharp/interop/qasm/_run.py

```diff
@@ -0,0 +1,171 @@
+          - target_profile (TargetProfile): The target profile to use for code generation.
+          - search_path (Optional[str]): The optional search path for resolving file references.
+          - output_semantics (OutputSemantics, optional): The output semantics for the compilation.
+          - seed (int): The seed to use for the random number generator.
+
+    Returns:
+        values: A list of results or runtime errors. If `save_events` is true,
+            a List of ShotResults is returned.
+
+    Raises:
+        QasmError: If there is an error generating, parsing, or analyzing the OpenQASM source.
+        QSharpError: If there is an error interpreting the input.
+        ValueError: If the number of shots is less than 1.
+        ValueError: If the `on_result` and `save_events` parameters are used when running OpenQASM programs.
+    """
+    warnings.warn(
+        "The 'run' API is experimental and subject to change in future releases.",
+        category=UserWarning,
+        stacklevel=2,
+    )
+
+    ipython_helper()
+
+    if shots < 1:
+        raise ValueError("The number of shots must be greater than 0.")
+
+    telemetry_events.on_run_qasm(shots)
+    start_time = monotonic()
+
+    results: List[ShotResult] = []
+
+    def on_save_events(output: Output) -> None:
+        # Append the output to the last shot's output list
+        results[-1]["events"].append(output)
+        if output.is_matrix():
+            results[-1]["matrices"].append(output)
+        elif output.is_state_dump():
+            results[-1]["dumps"].append(StateDump(output.state_dump()))
+        elif output.is_message():
+            results[-1]["messages"].append(str(output))
+
+    callable = None
+    if isinstance(source, Callable) and hasattr(source, "__global_callable"):
+        if len(args) == 1:
+            args = args[0]
+        elif len(args) == 0:
+            args = None
+        callable = source.__global_callable
+        source = None
+
+    if callable:
+        for _ in range(shots):
+            results.append(
+                {
+                    "result": None,
+                    "events": [],
+                    "matrices": [],
+                    "dumps": [],
+                    "messages": [],
+                }
+            )
+            run_results = get_interpreter().run(
+                source,
+                on_save_events if save_events else display_or_print,
+                noise,
+                callable=callable,
+                args=args,
+            )
+            results[-1]["result"] = run_results
+
+            if on_result:
+                on_result(results[-1])
+            # For every shot after the first, treat the entry expression as None to trigger
+            # a rerun of the last executed expression without paying the cost for any additional
+            # compilation.
+            source = None
+        if not save_events:
```

Comment:
Technically this isn't needed... we only go into this branch if `callable` is defined, and `callable` is only defined in the branch above where `source` is already set to `None`.

```suggestion
        if not save_events:
```